### PR TITLE
Adding ND support for tilize/untilize with padding

### PR DESCRIPTION
**Pull request by nardoTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/15935

### Problem description
Supporting ND tensors for tilize_with_val_padding and untilize_with_unpadding operations

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/12280234070
- [x] Blackhole Post commit (if applicable) https://github.com/tenstorrent/tt-metal/actions/runs/12280247340
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) tests passes
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp [hunk 1]
```diff
@@ -7,11 +7,78 @@
...
 
 namespace ttnn::operations::data_movement {
 
+using OwnedTilizeValArgs = std::tuple<ttnn::Tensor>;
+using BaseTilizeValType = std::function<ttnn::Tensor(const ttnn::Tensor&)>;
+
+using MassagedTilizeVal = MassagedOperation<ttnn::Tensor, const ttnn::Tensor&>;
+using MassagedTilizeValParams = MassagedOperationParams<ttnn::Tensor, const ttnn::Tensor&>;
+
+uint TILE_WIDTH = 32;
```
Comment (sjameelTT) [inline ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp:23]: use tt::constants::TILE_WIDTH and TILE_HEIGHT

File: tests/ttnn/unit_tests/operations/test_to_layout_5D.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+import ttnn
+import pytest
+from tests.ttnn.utils_for_testing import assert_with_pcc
+
+
```
Comment (ntarafdar) [inline tests/ttnn/unit_tests/operations/test_to_layout_5D.py:10]: Can we add 6,7,8 rank tensor tests

File: ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp [hunk 3]
```diff
@@ -7,11 +7,78 @@
...
+        std::array<uint32_t, 4> output_shape_4d;
+        output_shape_4d[0] = 1;
+        int extra_rank = output_shape.rank() - 4;
+        for (int i = extra_rank; i >= 0; i--) {
+            output_shape_4d[0] *= output_shape[i];
+        }
+        output_shape_4d[1] = output_shape[1 + extra_rank];
+        output_shape_4d[2] = output_shape[2 + extra_rank];
+        output_shape_4d[3] = output_shape[3 + extra_rank];
+        return tt::tt_metal::LegacyShape(output_shape_4d);
```
Comment (ntarafdar) [inline ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp:77]: Can we see if we can use Simple Shape? I think @jvegaTT did that recently, consult with him if you are having trouble.

File: ttnn/cpp/ttnn/operations/data_movement/untilize_with_unpadding/untilize_with_unpadding.cpp [hunk 4]
```diff
@@ -8,10 +8,53 @@
...
+        output_shape_4d[0] = 1;
+        int extra_rank = output_shape.rank() - 4;
+        for (int i = extra_rank; i >= 0; i--) {
+            output_shape_4d[0] *= (output_shape[i] + 1);
+        }
+        output_shape_4d[0]--;
+        output_shape_4d[1] = output_shape[1 + extra_rank];
+        output_shape_4d[2] = output_shape[2 + extra_rank];
+        output_shape_4d[3] = output_shape[3 + extra_rank];
+        return tt::tt_metal::LegacyShape(output_shape_4d);
```
Comment (ntarafdar) [inline ttnn/cpp/ttnn/operations/data_movement/untilize_with_unpadding/untilize_with_unpadding.cpp:28]: Same simple shape comment

**Review by ntarafdar — APPROVED** [review]
Just a few comments on testing and simple shape. Otherwise lgtm

File: ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp [hunk 5]
```diff
@@ -7,11 +7,78 @@
...
+using BaseTilizeValType = std::function<ttnn::Tensor(const ttnn::Tensor&)>;
+
+using MassagedTilizeVal = MassagedOperation<ttnn::Tensor, const ttnn::Tensor&>;
+using MassagedTilizeValParams = MassagedOperationParams<ttnn::Tensor, const ttnn::Tensor&>;
+
+uint TILE_WIDTH = 32;
+uint TILE_HEIGHT = 32;
+ttnn::Shape update_original_shape(ttnn::Shape& original) {
+    std::vector<uint32_t> update_original(original.rank());
+    if (original[original.rank() - 2] % TILE_HEIGHT != 0) {
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp:27]: Could you please save `original.rank() - 2]` and `original.rank() - 1]` into variables and use the variables instead.

File: ttnn/cpp/ttnn/operations/data_movement/untilize_with_unpadding/untilize_with_unpadding.cpp [hunk 6]
```diff
@@ -8,10 +8,53 @@
 #include "ttnn/common/constants.hpp"
 #include "ttnn/run_operation.hpp"
 
+#include "ttnn/operations/data_movement/common/common.hpp"
+#include "ttnn/operations/data_movement/reshape_view/reshape.hpp"
+
 using namespace tt::tt_metal;
 
+LegacyShape squeeze_output_shape(tt::tt_metal::LegacyShape output_shape) {
```
Comment (llongTT) [inline ttnn/cpp/ttnn/operations/data_movement/untilize_with_unpadding/untilize_with_unpadding.cpp:16]: I wonder if any generic method for arbitrary ND shape, as there will be requirement for higher dimension.

Comment (jaykru-tt) [reply ttnn/cpp/ttnn/operations/data_movement/untilize_with_unpadding/untilize_with_unpadding.cpp:16]: @llongTT This looks generic to me. 

But @nardoTT I did have a question about this function. You write `std::vector<uint32_t> output_shape_4d(output_shape.rank());` so the output_shape_4d has the same rank as the original rank when passed to LegacyShape. Should this instead be a 4-element vector? It looks like the later elements are never filled in. or does UntilizeWithUnpadding's run have some special interpretation of 0's in output_end?

**Review by jaykru-tt — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_to_layout_5D.py (hunk 2 above)
Comment (nardoTT) [reply tests/ttnn/unit_tests/operations/test_to_layout_5D.py:10]: Yes, I have added a similar test for 6D (7D, 8D, etc. are also similar). However for all the dimensions greater than 4, the value should still be 1 for tilize/untilize with or without padding. This is something we need to fix in the future.

File: ttnn/cpp/ttnn/operations/data_movement/untilize_with_unpadding/untilize_with_unpadding.cpp (hunk 6 above)
Comment (nardoTT) [reply ttnn/cpp/ttnn/operations/data_movement/untilize_with_unpadding/untilize_with_unpadding.cpp:16]: It is actually a 4-element vector because the output_end_vector that we are passing to squeeze_output_shape function is 4D. squeeze_output_shape function just updates the value of the first element in output_end_vector to account for the dimensions greater than 4.

**Review by bbradelTT — APPROVED** [review]
It'd be good to see if SimpleShape could be used instead of Shape as Naif pointed out. Besides that LGTM.

File: tests/ttnn/unit_tests/test_to_layout.py [hunk 7]
```diff
@@ -140,3 +140,27 @@ def test_to_layout_device(device, h, w, input_layout, output_layout):
     torch_brought_back = ttnn.to_torch(new_layout_tensor)
 
     assert_with_pcc(torch_input_tensor, torch_brought_back)
+
+
+@pytest.mark.parametrize("shape", [[1, 50, 1, 3, 768], [1, 1370, 1, 3, 1280]])
+@pytest.mark.parametrize("input_layout", [ttnn.TILE_LAYOUT, ttnn.ROW_MAJOR_LAYOUT])
+@pytest.mark.parametrize("output_layout", [ttnn.TILE_LAYOUT, ttnn.ROW_MAJOR_LAYOUT])
+def test_to_layout_5D(shape, input_layout, output_layout, device):
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/test_to_layout.py:148]: Should sweep across more dtypes for tests like this, especially `ttnn.bfloat8_b`. They go through different code paths.

File: ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp [hunk 8]
```diff
@@ -7,11 +7,78 @@
...
 
 namespace ttnn::operations::data_movement {
 
+using OwnedTilizeValArgs = std::tuple<ttnn::Tensor>;
+using BaseTilizeValType = std::function<ttnn::Tensor(const ttnn::Tensor&)>;
+
+using MassagedTilizeVal = MassagedOperation<ttnn::Tensor, const ttnn::Tensor&>;
+using MassagedTilizeValParams = MassagedOperationParams<ttnn::Tensor, const ttnn::Tensor&>;
+
+ttnn::Shape update_original_shape(ttnn::Shape& original) {
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp:23]: Use `SimpleShape`?

File: ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp [hunk 9]
```diff
@@ -7,11 +7,78 @@
...
+    std::vector<uint32_t> update_original(original.rank());
+    uint32_t indx1 = original.rank() - 1;
+    uint32_t indx2 = original.rank() - 2;
+    if (original[indx2] % tt::constants::TILE_HEIGHT != 0) {
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp:27]: Can you pass in tile instead of hardcoding constants?

File: ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp [hunk 10]
```diff
@@ -7,11 +7,78 @@
...
+        update_original[indx2] = (original[indx2] / tt::constants::TILE_HEIGHT + 1) * tt::constants::TILE_HEIGHT;
+        for (int i = 0; i < original.rank(); i++) {
+            if (i != indx2) {
+                update_original[i] = original[i];
+            }
+        }
+        return ttnn::Shape(tt::tt_metal::LegacyShape(update_original));
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp:34]: Even if using `ttnn::Shape`, this is unnecessary

File: ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp [hunk 11]
```diff
@@ -7,11 +7,78 @@
...
+            if (i != indx1) {
+                update_original[i] = original[i];
+            }
+        }
+        return ttnn::Shape(tt::tt_metal::LegacyShape(update_original));
+    }
+    return original;
+}
+
+MassagedTilizeVal build_ndiml_tilize_val(BaseTilizeValType base_tilize) {
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp:49]: `ndiml`?

File: ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp [hunk 12]
```diff
@@ -7,11 +7,78 @@
...
+    auto original_shape = std::make_shared<ttnn::Shape>(ttnn::Shape{});
+    return MassagedTilizeVal(MassagedTilizeValParams{
+        .predicate = [](const ttnn::Tensor& input_tensor) -> bool { return input_tensor.get_shape().rank() > 4; },
+        .pre_transform = [=](const ttnn::Tensor& input_tensor) -> OwnedTilizeValArgs {
+            *original_shape = input_tensor.get_shape();
+            ttnn::Tensor squeezed_tensor = squeeze_to_le_4D(input_tensor);
+            auto print_shape = squeezed_tensor.get_shape();
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp:56]: Remove?

File: ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp [hunk 13]
```diff
@@ -7,11 +7,78 @@
...
+                update_original[i] = original[i];
+            }
+        }
+        return ttnn::Shape(tt::tt_metal::LegacyShape(update_original));
+    }
+    return original;
+}
+
+MassagedTilizeVal build_ndiml_tilize_val(BaseTilizeValType base_tilize) {
+    auto original_shape = std::make_shared<ttnn::Shape>(ttnn::Shape{});
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp:50]: Is this really necessary?

File: ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp [hunk 14]
```diff
@@ -7,11 +7,78 @@
...
+    return MassagedTilizeVal(MassagedTilizeValParams{
+        .predicate = [](const ttnn::Tensor& input_tensor) -> bool { return input_tensor.get_shape().rank() > 4; },
+        .pre_transform = [=](const ttnn::Tensor& input_tensor) -> OwnedTilizeValArgs {
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp:53]: Is there a reason to use copy capture by default?

File: ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp [hunk 15]
```diff
@@ -54,8 +124,8 @@ ttnn::Tensor ExecuteTilizeWithZeroPadding::invoke(
     using namespace tt::constants;
     auto shape = input_tensor.get_legacy_shape();
 
-    shape[2] = tt::round_up(shape[2], TILE_HEIGHT);
-    shape[3] = tt::round_up(shape[3], TILE_WIDTH);
+    shape[2] = tt::round_up(shape[2], tt::constants::TILE_HEIGHT);
+    shape[3] = tt::round_up(shape[3], tt::constants::TILE_WIDTH);
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp:128]: Probably should start thinking about tiny tiles :D

**Review by TT-BrianLiu — APPROVED** [review]
Overall code seems ok, but please switch to `SimpleShape` where possible. These ops are probably most sensitive to issues with `LegacyShape`.

Will approve, but please address comments.

File: ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp (hunk 8 above)
Comment (nardoTT) [reply ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp]: we are working on SimpleShape and other data types in other PRs cause this one is currently needed for other operations

File: ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp (hunk 11 above)
Comment (nardoTT) [reply ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.cpp:49]: it is n-dimensional
